### PR TITLE
Feature/Guardian on Nodes + Initial OSF Groups Modeling [PLAT-1081][PLAT-1176]

### PR DESCRIPTION
**Pull request by pattisdr**
## Purpose

Change our permissions system to use django-guardian (add the groups and permissions to the Node model, add GuardianMixin, etc).  Also add initial modeling changes for Groups feature.  Groups feature is not yet exposed, but queries have been rewritten to accommodate them.

## Changes
- Adds GuardianMixin to AbstractNode model
- On node creation, creates an admin group, write group, and read group for the node.  Admin group gets admin-write-read perms, write group gets write-read, and read group gets read perms. Creator automatically added to node's admin group.
- Consolidates contributor methods in ContributorMixin to now work for both nodes and preprints since they both use guardian
- Adds data migration to create django groups for existing nodes and add contributors to them.
- Schema migration removes admin/write/read fields from Contributor table
- Migration added (by Casey) which uses direct foreign keys w/ guardian
- Changes to permission-related queries to use guardian
- Does NOT explicitly add admin contributors as read contributors to other groups - okayed at Engineering mtg.  This gets too messy, and adds way too much code clutter.  When you add contributors, add components, remove contributors, update contributors, remove projects, you have to do all this extra checking for who needs to be added/removed as an admin contributor.  Also, we'd have to have another table to keep track of who was added where, were you explicitly added or implicitly added? With those extra lookups, I think it makes sense to just query parents to see if any admins are in the tree and optimize as much as possible.
- Adds OSFGroups model.  On creation, automatically creates a member and manager group, member has member permissions and manager has manager permissions.  Methods added to manage group membership.
- Methods added to Node model to add/remove/update groups on node.
- Queries rewritten so sometimes checking for permission instead of contributorship, except in cases where we specifically care if a user is an explicit contributor
- Method added to user model to fetch all user groups

## For the Reviewer
- Be thinking about the places where the distinction between whether a user is a traditional contributor or group member are important.
- Not sure how efficient some of these guardian shortcuts are. Are there places where we need to optimize these new queries?
- In general, we are trying to *match* implicit admin behavior.  UserNodes, for example, don't include nodes for which you have implicit admin, but NodeChildren does. Places on the project overview page look different for implicit admin versus traditional contributors.  Per product, they are fine with matching behavior.

## QA Notes

There should be no user-facing changes - but this has the potential to break permission-related things. How users get permissions and how permissions are checked have been changed. OSF Groups exist, but this functionality is not yet exposed.  Would do a thorough regression of the project page, editing project, files, wiki, registrations, settings, contributors, especially permission-related things.  It should work as before.

## Ticket

https://openscience.atlassian.net/browse/PLAT-1081

### CODE REVIEW

**Comment by pattisdr** [comment]
I think this is ready to go on feature/guardian @caseyrollins !

File: api/base/utils.py [hunk 1]
```diff
@@ -151,8 +152,8 @@ def default_node_permission_queryset(user, model_cls):
     assert model_cls in {Node, Registration}
     if user.is_anonymous:
         return model_cls.objects.filter(is_public=True)
-    sub_qs = Contributor.objects.filter(node=OuterRef('pk'), user__id=user.id, read=True)
-    return model_cls.objects.annotate(contrib=Exists(sub_qs)).filter(Q(contrib=True) | Q(is_public=True))
+    read_user_query = Q(id__in=get_objects_for_user(user, 'read_node', model_cls))
```
Comment (sloria) [inline api/base/utils.py]: Some drive-by nits:

* Would be good to pass `with_superuser=False`
* You can use `values_list('id')` here

File: api/nodes/filters.py [hunk 2]
```diff
@@ -78,16 +79,16 @@ def build_query_from_field(self, field_name, operation):
 class UserNodesFilterMixin(NodesFilterMixin):
     def build_query_from_field(self, field_name, operation):
         if field_name == 'current_user_permissions':
-            if not operation['value'] in PERMISSIONS:
+            if not operation['value'] in API_CONTRIBUTOR_PERMISSIONS:
```
Comment (sloria) [inline api/nodes/filters.py]: Drive-by nit: Prefer `if x not in y:` to yoda conditions.

I realize this is how it was before, though.

File: api/nodes/filters.py [hunk 3]
```diff
@@ -78,16 +79,16 @@ def build_query_from_field(self, field_name, operation):
...
                 raise InvalidFilterValue(value=operation['value'])
             perm = operation['value']
-            query = Q(contributor__user__id=self.get_user().id)
+            user = self.get_user()
             if perm == READ:
-                query &= Q(contributor__read=True)
+                query = Q(id__in=get_objects_for_user(user, 'read_node', AbstractNode))
```
Comment (sloria) [inline api/nodes/filters.py]: Same with these: pass `with_superuser=False` and use `.values_list('id')`.

**Comment by pattisdr** [comment]
Thank you @sloria! I'm creating a branch to bring feature/guardian up-to-date and I'll add those changes plus a few we discussed in our in-person CR there.

**Comment by pattisdr** [comment]
Comments addressed here https://github.com/CenterForOpenScience/osf.io/pull/8834
